### PR TITLE
Fix for race condition when receiving logon response before logon request has been processed by MessageStore

### PR DESCRIPTION
**Pull request by jonfreedman**
As discussed on the [mailing list](https://sourceforge.net/p/quickfixj/mailman/quickfixj-users/thread/1974CA77F2BC6F4AA91619016DCFC9EF7DE94FD7%40P2EXCH01.rivagecapital.com) there appears to be a race condition between storage of logon request and receipt of logon response when validating the expected SeqNum.  `SessionDisconnectTest` simulates this scenario and locking around the call to `SessionState#getNextSenderMsgSeqNum` in `Session#nextLogon` fixes the specific issue.

### CODE REVIEW

**Comment by chrjohn** [comment]
Thanks for the PR. The Builder looks nice. Have to check how I can merge the tests to 1.6.x which is still on JDK7. ;)
Cheers

File: quickfixj-core/src/main/java/quickfix/Session.java [hunk 1]
```diff
@@ -2057,7 +2057,9 @@ private void nextLogon(Message logon) throws FieldNotFound, RejectLogon, Incorre
         if (logon.isSetField(NextExpectedMsgSeqNum.FIELD) && enableNextExpectedMsgSeqNum) {
 
             final int targetWantsNextSeqNumToBe = logon.getInt(NextExpectedMsgSeqNum.FIELD);
-            final int actualNextNum = state.getMessageStore().getNextSenderMsgSeqNum();
+            state.lockSenderMsgSeqNum();
+            final int actualNextNum = state.getNextSenderMsgSeqNum();
```
Comment (chrjohn) [inline quickfixj-core/src/main/java/quickfix/Session.java]: Could you please wrap this in a try-finally-block? Thanks.

Comment (jonfreedman) [inline quickfixj-core/src/main/java/quickfix/Session.java]: Sorry, I realised on my way out the door last night I forgot this

**Comment by jonfreedman** [comment]
Regarding JDK7 it appears the only module with tests is quickfixj-core so you could introduce a new quickfixj-tests module which is compiled using JDK8?

**Pull request merged**